cirq/linalg/transformations.py

```diff
@@ -304,3 +305,47 @@ def partial_trace(tensor: np.ndarray,
     left_indices = [keep_map[i] if i in keep_set else i for i in range(ndim)]
     right_indices = [ndim + i if i in keep_set else i for i in left_indices]
     return np.einsum(tensor, left_indices + right_indices)
+
+
+def keep_qubits(state: np.ndarray, keep_indices: List[int]) -> np.ndarray:
+    """Return a representation of the given state over a subset of its qubits.
+
+    A given wavefunction can be expressed over a subset of its qubits by
+    tracing out undesired qubits from the state's corresponding density matrix.
+    If the traced density matrix is rank 1 (pure state), then it can be
+    refactored into wavefunction form via eigendecomposition. Otherwise, the
+    outcome is a mixed state and cannot be expressed as a wavefunction.
+
+    Global phase of the input state will not necessariy be preserved in the
+    output state.
+
+    Args:
+        state: A wavefunction-like
+
+    """
+
+    if not np.isclose(np.linalg.norm(state), 1):
+        raise ValueError("Input state must be normalized.")
+    if not len(set(keep_indices)) == len(keep_indices):
+        raise ValueError(
+            "keep_indices were {} but must be unique.".format(keep_indices))
+    dim = np.log2(state.shape[0])
+    if not dim.is_integer():
+        raise ValueError("Input wavefunction has bad length of {}.".format(
+            state.shape[0]))
+    rho = np.kron(np.conj(np.reshape(state, (state.shape[0], 1)).T),
+                  state).reshape((2, 2) * int(dim))
+    new_dim = 2**len(keep_indices)
+    keep_rho = partial_trace(rho, keep_indices).reshape((new_dim,) * 2)
+    if not any(keep_indices):
+        return keep_rho
+
+    # FIXME: reasonable tolerance for purity check?
+    if approx_eq(np.trace(keep_rho @ keep_rho), 1):
+        w, v = np.linalg.eig(keep_rho)
+        # FIXME: reasonable tolerance for lone eigenvalue check?
+        return v[:, np.isclose(w, 1)].reshape((new_dim,))
+
+    # FIXME: not sure how to deal with warnings
+    print("Removing qubits from an entangled state resulted in mixture.")
```

Comment:
The caller is either going to consider this to be an error, or they are going to want a sampled result from the mixture, or they are going to want the mixture (or density matrix or similar).

You should strongly distinguish between the use case where the caller expects the result to be pure, and the use case where the caller does not known. In fact these should be completely different methods with different names. One of them returns the single state or else fails, and the other always returns a mixture (which the caller can then inspect to see if it has length 1).

Reply:
Will do.